diff --git a/vllm/model_executor/models/deepseek_v2.py b/vllm/model_executor/models/deepseek_v2.py
index f1c4a7b21..9e4f87105 100644
--- a/vllm/model_executor/models/deepseek_v2.py
+++ b/vllm/model_executor/models/deepseek_v2.py
@@ -938,7 +938,7 @@ class DeepseekV2MLAAttention(nn.Module):
             mscale = yarn_get_mscale(scaling_factor, float(mscale_all_dim))
             self.scaling = self.scaling * mscale * mscale
 
-        self.is_v32 = hasattr(config, "index_topk")
+        self.is_v32 = bool(getattr(config, "index_topk", None))
 
         if self.is_v32:
             self.indexer_rope_emb = get_rope(
@@ -1152,14 +1152,15 @@ class DeepseekV2Model(nn.Module):
         self.device = current_platform.device_type
 
         self.vocab_size = config.vocab_size
-        self.is_v32 = hasattr(config, "index_topk")
+        self.is_v32 = bool(getattr(config, "index_topk", None))
         if self.is_v32:
             topk_tokens = config.index_topk
+            _device = "cpu" if str(self.device) == "tpu" else self.device
             topk_indices_buffer = torch.empty(
                 vllm_config.scheduler_config.max_num_batched_tokens,
                 topk_tokens,
                 dtype=torch.int32,
-                device=self.device,
+                device=_device,
             )
         else:
             topk_indices_buffer = None
@@ -1464,6 +1465,8 @@ class DeepseekV2ForCausalLM(
         for name, loaded_weight in weights:
             if "rotary_emb.inv_freq" in name:
                 continue
+            if "indexer" in name and not bool(getattr(self.config, "index_topk", None)):
+                continue
 
             spec_layer = get_spec_layer_idx_from_weight_name(self.config, name)
             if spec_layer is not None:
